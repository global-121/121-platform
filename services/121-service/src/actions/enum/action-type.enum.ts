export enum ActionType {
  startBlockNewPayment = 'start-block-new-payment',
  endBlockNewPayment = 'end-block-new-payment',
}
