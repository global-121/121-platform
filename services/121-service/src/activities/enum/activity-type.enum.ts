export enum ActivityTypeEnum {
  DataChange = 'data-change',
  FinancialServiceProviderChange = 'financial-service-provider-change',
  Message = 'message',
  Note = 'note',
  StatusChange = 'status-change',
  Transaction = 'transaction',
}
