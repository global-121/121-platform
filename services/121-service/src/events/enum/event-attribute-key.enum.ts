export enum EventAttributeKeyEnum {
  reason = 'reason',
  oldValue = 'oldValue',
  newValue = 'newValue',
  fieldName = 'fieldName',
  duplicateWithRegistrationId = 'duplicateWithRegistrationId',
  duplicateWithRegistrationProgramId = 'duplicateWithRegistrationProgramId',
}
