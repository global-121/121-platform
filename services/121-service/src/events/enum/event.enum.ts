export enum EventEnum {
  registrationDataChange = 'registrationDataChange',
  financialServiceProviderChange = 'financialServiceProviderChange',
  registrationStatusChange = 'registrationStatusChange',
  ignoredDuplicate = 'ignoredDuplicate',
}
