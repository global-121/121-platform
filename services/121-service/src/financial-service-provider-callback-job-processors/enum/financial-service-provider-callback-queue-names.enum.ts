export enum FinancialServiceProviderCallbackQueueNames {
  safaricomTransferCallback = 'safaricomTransferCallback',
  safaricomTimeoutCallback = 'safaricomTimeoutCallback',
}
