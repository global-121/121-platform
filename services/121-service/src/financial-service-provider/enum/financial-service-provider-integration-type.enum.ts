export enum FinancialServiceProviderIntegrationType {
  api = 'api',
  csv = 'csv',
  xml = 'xml',
}
