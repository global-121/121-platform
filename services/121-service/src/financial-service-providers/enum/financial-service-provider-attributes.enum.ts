export enum FinancialServiceProviderAttributes {
  phoneNumber = 'phoneNumber',
  nationalId = 'nationalId',
  fullName = 'fullName',
  addressStreet = 'addressStreet',
  addressHouseNumber = 'addressHouseNumber',
  addressHouseNumberAddition = 'addressHouseNumberAddition',
  addressPostalCode = 'addressPostalCode',
  addressCity = 'addressCity',
  bankAccountNumber = 'bankAccountNumber',
  whatsappPhoneNumber = 'whatsappPhoneNumber',
}
