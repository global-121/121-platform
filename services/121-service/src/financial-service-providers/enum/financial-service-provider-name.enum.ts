export enum FinancialServiceProviderName {
  intersolveVoucherWhatsapp = 'Intersolve-voucher-whatsapp',
  intersolveVoucherPaper = 'Intersolve-voucher-paper',
  intersolveVisa = 'Intersolve-visa',
  intersolveJumboPhysical = 'Intersolve-jumbo-physical',
  africasTalking = 'Africas-talking',
  belcash = 'BelCash',
  vodacash = 'VodaCash',
  bobFinance = 'BoB-finance',
  ukrPoshta = 'UkrPoshta',
  safaricom = 'Safaricom',
  commercialBankEthiopia = 'Commercial-bank-ethiopia',
  excel = 'Excel',
}

export enum FinancialServiceProviderConfigurationEnum {
  password = 'password',
  username = 'username',
  columnsToExport = 'columnsToExport',
  columnToMatch = 'columnToMatch',
  brandCode = 'brandCode',
  displayName = 'displayName',
  coverLetterCode = 'coverLetterCode',
}

export const FinancialServiceProviderConfigurationMapping: {
  [key in FinancialServiceProviderName]?: any;
} = {
  [FinancialServiceProviderName.intersolveVoucherWhatsapp]: [
    FinancialServiceProviderConfigurationEnum.password,
    FinancialServiceProviderConfigurationEnum.username,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.intersolveVoucherPaper]: [
    FinancialServiceProviderConfigurationEnum.password,
    FinancialServiceProviderConfigurationEnum.username,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.intersolveVisa]: [
    FinancialServiceProviderConfigurationEnum.brandCode,
    FinancialServiceProviderConfigurationEnum.coverLetterCode,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.intersolveJumboPhysical]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.africasTalking]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.belcash]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.vodacash]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.bobFinance]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.ukrPoshta]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.safaricom]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.commercialBankEthiopia]: [
    FinancialServiceProviderConfigurationEnum.password,
    FinancialServiceProviderConfigurationEnum.username,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.excel]: [
    FinancialServiceProviderConfigurationEnum.columnsToExport,
    FinancialServiceProviderConfigurationEnum.columnToMatch,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
};
