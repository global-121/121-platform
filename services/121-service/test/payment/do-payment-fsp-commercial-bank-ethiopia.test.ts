import { HttpStatus } from '@nestjs/common';

import { TransactionStatusEnum } from '@121-service/src/payments/transactions/enums/transaction-status.enum';
import { SeedScript } from '@121-service/src/scripts/enum/seed-script.enum';
import {
  doPayment,
  getTransactions,
  retryPayment,
  waitForPaymentTransactionsToComplete,
} from '@121-service/test/helpers/program.helper';
import { seedIncludedRegistrations } from '@121-service/test/helpers/registration.helper';
import {
  getAccessToken,
  resetDB,
} from '@121-service/test/helpers/utility.helper';
import { registrationCbe } from '@121-service/test/registrations/pagination/pagination-data';

const programId = 1;
const payment = 1;
const amount = 200;

describe('Do payment with FSP: Commercial Bank of Ethiopia', () => {
  let accessToken: string;

  beforeEach(async () => {
    await resetDB(SeedScript.cbeProgram);
    accessToken = await getAccessToken();
  });

  it('when credit transfer API call gives a success response should successfully do a payment', async () => {
    // Arrange
    const paymentReferenceIds = [registrationCbe.referenceId];
    await seedIncludedRegistrations([registrationCbe], programId, accessToken);

    // Act
    const doPaymentResponse = await doPayment({
      programId,
      paymentNr: payment,
      amount,
      referenceIds: paymentReferenceIds,
      accessToken,
    });

    await waitForPaymentTransactionsToComplete({
      programId,
      paymentReferenceIds,
      accessToken,
      maxWaitTimeMs: 4000,
      completeStatusses: [
        TransactionStatusEnum.success,
        TransactionStatusEnum.error,
      ],
    });

    // Assert
    const getTransactionsBody = await getTransactions({
      programId,
      paymentNr: payment,
      referenceId: registrationCbe.referenceId,
      accessToken,
    });

    expect(doPaymentResponse.status).toBe(HttpStatus.ACCEPTED);
    expect(doPaymentResponse.body.applicableCount).toBe(
      paymentReferenceIds.length,
    );
    expect(getTransactionsBody.body[0].status).toBe(
      TransactionStatusEnum.success,
    );
    expect(getTransactionsBody.body[0].errorMessage).toBe(null);
  });

  it('when credit transfer API call gives an error response should succesfully do a payment with transactions that have status error', async () => {
    // Arrange
    const paymentReferenceIds = [registrationCbe.referenceId];
    // The fullName value triggers a specific mock scenario
    const registrationCbeWithError = {
      ...registrationCbe,
      fullName: 'error',
    };

    await seedIncludedRegistrations(
      [registrationCbeWithError],
      programId,
      accessToken,
    );

    // Act
    const doPaymentResponse = await doPayment({
      programId,
      paymentNr: payment,
      amount,
      referenceIds: paymentReferenceIds,
      accessToken,
    });

    await waitForPaymentTransactionsToComplete({
      programId,
      paymentReferenceIds,
      accessToken,
      maxWaitTimeMs: 4000,
      completeStatusses: [
        TransactionStatusEnum.success,
        TransactionStatusEnum.error,
      ],
    });

    // Assert
    const getTransactionsBody = await getTransactions({
      programId,
      paymentNr: payment,
      referenceId: registrationCbeWithError.referenceId,
      accessToken,
    });

    expect(doPaymentResponse.status).toBe(HttpStatus.ACCEPTED);
    expect(doPaymentResponse.body.applicableCount).toBe(
      paymentReferenceIds.length,
    );
    expect(getTransactionsBody.body[0].status).toBe(
      TransactionStatusEnum.error,
    );
    expect(getTransactionsBody.body[0].errorMessage).toMatchSnapshot();
  });

  it('when credit transfer API call times out should succesfully do a payment with transactions that have status error', async () => {
    // Arrange
    const paymentReferenceIds = [registrationCbe.referenceId];

    // The fullName value triggers a specific mock scenario
    const registrationCbeWithTimeout = {
      ...registrationCbe,
      fullName: 'time-out',
    };

    await seedIncludedRegistrations(
      [registrationCbeWithTimeout],
      programId,
      accessToken,
    );

    // Act
    const doPaymentResponse = await doPayment({
      programId,
      paymentNr: payment,
      amount,
      referenceIds: paymentReferenceIds,
      accessToken,
    });

    await waitForPaymentTransactionsToComplete({
      programId,
      paymentReferenceIds,
      accessToken,
      maxWaitTimeMs: 4000,
      completeStatusses: [
        TransactionStatusEnum.success,
        TransactionStatusEnum.error,
      ],
    });

    // Assert
    const getTransactionsBody = await getTransactions({
      programId,
      paymentNr: payment,
      referenceId: registrationCbeWithTimeout.referenceId,
      accessToken,
    });

    expect(doPaymentResponse.status).toBe(HttpStatus.ACCEPTED);
    expect(doPaymentResponse.body.applicableCount).toBe(
      paymentReferenceIds.length,
    );
    expect(getTransactionsBody.body[0].status).toBe(
      TransactionStatusEnum.error,
    );
    expect(getTransactionsBody.body[0].errorMessage).toMatchSnapshot();
  });

  it('when credit transfer API call returns duplicated transaction error should succesfully do a retry payment with transactions that have status success when transaction enquiry returns a success response', async () => {
    // Arrange
    const paymentReferenceIds = [registrationCbe.referenceId];

    // The fullName value triggers a specific mock scenario
    const registrationCbeWithTimeout = {
      ...registrationCbe,
      fullName: 'time-out',
    };

    await seedIncludedRegistrations(
      [registrationCbeWithTimeout],
      programId,
      accessToken,
    );

    const doPaymentResponse = await doPayment({
      programId,
      paymentNr: payment,
      amount,
      referenceIds: paymentReferenceIds,
      accessToken,
    });

    await waitForPaymentTransactionsToComplete({
      programId,
      paymentReferenceIds,
      accessToken,
      maxWaitTimeMs: 4000,
      completeStatusses: [
        TransactionStatusEnum.success,
        TransactionStatusEnum.error,
      ],
    });

    const getTransactionsBody = await getTransactions({
      programId,
      paymentNr: payment,
      referenceId: registrationCbeWithTimeout.referenceId,
      accessToken,
    });

    expect(doPaymentResponse.status).toBe(HttpStatus.ACCEPTED);
    expect(doPaymentResponse.body.applicableCount).toBe(
      paymentReferenceIds.length,
    );
    expect(getTransactionsBody.body[0].status).toBe(
      TransactionStatusEnum.error,
    );

    // Act
    const doPaymentRetryResponse = await retryPayment({
      programId,
      paymentNr: payment,
      accessToken,
    });

    await waitForPaymentTransactionsToComplete({
      programId,
      paymentReferenceIds,
      accessToken,
      maxWaitTimeMs: 4000,
      completeStatusses: [TransactionStatusEnum.success],
    });

    const getTransactionsAfterRetryBody = await getTransactions({
      programId,
      paymentNr: payment,
      referenceId: registrationCbeWithTimeout.referenceId,
      accessToken,
    });

    // Assert
    // Check if the transaction status is success.
    expect(doPaymentRetryResponse.status).toBe(HttpStatus.OK);
    expect(doPaymentRetryResponse.body.applicableCount).toBe(
      paymentReferenceIds.length,
    );
    expect(getTransactionsAfterRetryBody.body[0].status).toBe(
      TransactionStatusEnum.success,
    );
    expect(getTransactionsAfterRetryBody.body[0].errorMessage).toBe(null);

    // TODO Implement the following test cases if we refactor CBE integration:
    // - should succesfully do a payment with transactions that have status error when transaction enquiry returns an error response
    // - should succesfully do a payment with transactions that have status error when transaction enquiry times out
  });
});
