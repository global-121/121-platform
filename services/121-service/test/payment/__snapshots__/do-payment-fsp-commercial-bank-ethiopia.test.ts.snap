// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Do payment with FSP: Commercial Bank of Ethiopia when credit transfer API call gives an error response should succesfully do a payment with transactions that have status error 1`] = `"Other failure"`;

exports[`Do payment with FSP: Commercial Bank of Ethiopia when credit transfer API call times out should succesfully do a payment with transactions that have status error 1`] = `"Failed because of an unknown error. Please contact 121 technical support."`;
